File: php/features.php
<?php
declare(strict_types=1);

// CustomerioPipelines SDK feature factory

require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/feature/DebugFeature.php';
require_once __DIR__ . '/feature/IdempotencyFeature.php';
require_once __DIR__ . '/feature/MetricsFeature.php';
require_once __DIR__ . '/feature/PagingFeature.php';
require_once __DIR__ . '/feature/RatelimitFeature.php';
require_once __DIR__ . '/feature/RetryFeature.php';
require_once __DIR__ . '/feature/TestFeature.php';
require_once __DIR__ . '/feature/TimeoutFeature.php';


class CustomerioPipelinesFeatures
{
    public static function make_feature(string $name)
    {
        switch ($name) {
            case "base":
                return new CustomerioPipelinesBaseFeature();
            case "debug":
                return new CustomerioPipelinesDebugFeature();
            case "idempotency":
                return new CustomerioPipelinesIdempotencyFeature();
            case "metrics":
                return new CustomerioPipelinesMetricsFeature();
            case "paging":
                return new CustomerioPipelinesPagingFeature();
            case "ratelimit":
                return new CustomerioPipelinesRatelimitFeature();
            case "retry":
                return new CustomerioPipelinesRetryFeature();
            case "test":
                return new CustomerioPipelinesTestFeature();
            case "timeout":
                return new CustomerioPipelinesTimeoutFeature();
            default:
                return new CustomerioPipelinesBaseFeature();
        }
    }

    /**
     * Does a generated feature class back this name? False for a name only
     * an options extend instance can supply (the station adopt path) - the
     * constructor uses this to skip make_feature for such names instead of
     * adding a stray BaseFeature.
     */
    public static function has_feature(string $name): bool
    {
        switch ($name) {
            case "base":
            case "debug":
            case "idempotency":
            case "metrics":
            case "paging":
            case "ratelimit":
            case "retry":
            case "test":
            case "timeout":
                return true;
            default:
                return false;
        }
    }
}

File: py/customeriopipelines_sdk/features.py
# CustomerioPipelines SDK feature factory

from customeriopipelines_sdk.feature.base_feature import CustomerioPipelinesBaseFeature
from customeriopipelines_sdk.feature.debug_feature import CustomerioPipelinesDebugFeature
from customeriopipelines_sdk.feature.idempotency_feature import CustomerioPipelinesIdempotencyFeature
from customeriopipelines_sdk.feature.metrics_feature import CustomerioPipelinesMetricsFeature
from customeriopipelines_sdk.feature.paging_feature import CustomerioPipelinesPagingFeature
from customeriopipelines_sdk.feature.ratelimit_feature import CustomerioPipelinesRatelimitFeature
from customeriopipelines_sdk.feature.retry_feature import CustomerioPipelinesRetryFeature
from customeriopipelines_sdk.feature.test_feature import CustomerioPipelinesTestFeature
from customeriopipelines_sdk.feature.timeout_feature import CustomerioPipelinesTimeoutFeature


_FEATURES = {
    "base": lambda: CustomerioPipelinesBaseFeature(),
    "debug": lambda: CustomerioPipelinesDebugFeature(),
    "idempotency": lambda: CustomerioPipelinesIdempotencyFeature(),
    "metrics": lambda: CustomerioPipelinesMetricsFeature(),
    "paging": lambda: CustomerioPipelinesPagingFeature(),
    "ratelimit": lambda: CustomerioPipelinesRatelimitFeature(),
    "retry": lambda: CustomerioPipelinesRetryFeature(),
    "test": lambda: CustomerioPipelinesTestFeature(),
    "timeout": lambda: CustomerioPipelinesTimeoutFeature(),
}


def _make_feature(name):
    factory = _FEATURES.get(name)
    if factory is not None:
        return factory()
    return _FEATURES["base"]()


# True when this SDK was generated with the named feature class - the
# constructor's tolerance for extend-carried features reads this (an
# active name with no generated class must not become a BaseFeature
# stray when an extend instance carries it).
def _has_feature(name):
    return name in _FEATURES
